### DISCUSSION TITLE
CD changer emulation (Nakamichi et al.)

### DISCUSSION THREAD

**Discussion by elosha** [BlueSCSI/BlueSCSI-v2#139] [Ideas]
Hey there, I am well aware that there is no LUN support yet. But I don't know the exact circumstances ;)

There is one device that gives LUNs a very practical twist and it would be really cool if emulation was possible: The CD-ROM changer. For example, look up Nakamichi drives.

Especially servers could profit, as they can mount multiple volumes of a multi-disc collection at once – even without the annoying changing delay of original hardware. All that with just one ID, assigning each disk to a LUN and thus conserving valuable IDs for use with HDDs and physical drives.

**Comment by zefrenchtoon** [comment]
May I add this link to show how it works on a Mac ?
https://stuffandnonsense.elephantandchicken.co.uk/?p=1603

**Comment by elosha** [comment]
Well, it should not be too hard to add LUN support, maybe I give it a try. Let's brainstorm …

I'll call the Nakamichi style changer a **multi-changer**.
I'll call the existing image switching mechanism (eject -> load next) a **single-changer**.

### Config Interface
I heard adding more options to bluescsi.ini is not popular. But we don't need to!

1. The CD single-changer emulation needs to remain intact, i.e. by filling a directory `CD5`, which feeds images into a CD-ROM drive on ID 5.
2. The file name definition for LUNs aleady exists, i.e. `CD50 abc.iso` + `CD51 def.iso` + `CD52 ghi.iso` should result in creating a CD ROM multi-changer on ID 5 containing LUNs 0 to 2, each supplied by an ISO image.
3. Regarding the bluescsi.ini, I could imagine dedicating the drive option `IMG0…n` to assigning images to LUNs.
4. The bluescsi.ini `DIR` device option would remain as the image option for the simpler single-changer interface.

**Reply by zefrenchtoon** [reply]
Here is another example reviewed by Pierre Dandumont on a Pioneer device and the direvers and tools seem interesting:

https://www.journaldulapin.com/2018/06/20/chargeur-jukebox-pioneer/
